zilliqa/src/precompiles/scilla.rs

```diff
@@ -333,24 +333,61 @@ pub fn scilla_call_handle_register<I: Inspector<PendingState> + ScillaInspector>
     // Call handler
     let prev_handle = handler.execution.call.clone();
     handler.execution.call = Arc::new(move |ctx, inputs| {
+        let _gas_limit = ctx.evm.db.gas_left.remaining();
         if inputs.bytecode_address != Address::from(*b"\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0ZIL\x53") {
             return prev_handle(ctx, inputs);
         }
 
-        let gas = Gas::new(inputs.gas_limit);
+        let zq1_gas_rule = {
+            let account = ctx.evm.db.load_account(inputs.caller).unwrap();
+            matches!(account.account.created_at_block, CreatedAtBlock::ZQ1)
+        };
+
+        let gas_limit = {
+            if zq1_gas_rule {
+                let total_gas_left = ctx.evm.db.gas_left.remaining();
+                let gas_cost = EvmGas::from(SCILLA_INVOKE_RUNNER).0
+                    + EvmGas::from(ScillaGas(ZIL_CONTRACT_INVOKE_GAS as u64)).0;
+                match total_gas_left.checked_sub(gas_cost) {
+                    Some(result) => result,
+                    _ => {
+                        return Ok(FrameOrResult::new_call_result(
+                            InterpreterResult {
+                                result: InstructionResult::PrecompileOOG,
+                                output: Bytes::new(),
+                                gas: Default::default(),
+                            },
+                            inputs.return_memory_offset.clone(),
+                        ))
+                    }
+                }
+            } else {
+                inputs.gas_limit
+            }
+        };
+
+        let gas = Gas::new(gas_limit);
         let outcome =
             scilla_call_precompile(&inputs, gas.limit(), &mut ctx.evm.inner, &mut ctx.external);
 
+        let result_gas = {
+            if zq1_gas_rule {
```

Comment:
Are you sure the ZQ1 implementation charges 0 gas? I thought it just charged less gas?